Cs/sc 1945 ability to add filter to rules

Charl1996 opened the pull request:
## Product Description
This PR adds an additional section to the deduplication create/edit [pages](https://staging.commcarehq.org/a/charlsmit-testy/data/edit/deduplication_rules/add/), namely "Case Filters", which allows a user to add filters to the deduplication rule, such that the rule is only applicable to cases matching the specified filter criteria.


## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-1945)

In addition to the new "Case Filters" section that has been added, a new location related criterion ([`LocationFilterDefinition`](https://github.com/dimagi/commcare-hq/blob/6bfbad00502f6e0800f913492e61fb398e39bd72/corehq/apps/data_interfaces/models.py#L673)) has been added to the existing `CaseRuleCriteriaForm` (see [here](https://github.com/dimagi/commcare-hq/blob/6a5d5f1a6d8a48f6a8e2bd28eb8f78f391046bde/corehq/apps/data_interfaces/forms.py#L281)) to allow a user to filter cases that belong to a certain location or is in a location's descendants.

### Adding crispy-ness
The new dedupe "Case Filters" section is modelled after the "Case Filters" (or really the "Rule Criteria" part) section seen in the [Automatically Update Cases](https://staging.commcarehq.org/a/charlsmit-testy/data/edit/automatic_updates/add/) feature, which uses the [CaseRuleCriteriaForm](https://github.com/dimagi/commcare-hq/blob/6a5d5f1a6d8a48f6a8e2bd28eb8f78f391046bde/corehq/apps/data_interfaces/forms.py#L266) crispy form. 
A new crispy form, [DedupeCaseFilterForm](https://github.com/dimagi/commcare-hq/blob/c89a503942614bc93062bce9558ab2a3968bd992/corehq/apps/data_interfaces/forms.py#L859) (inherits from `CaseRuleCriteriaForm`), was created to suite the dedupe feature's needs specifically (similar to what was done [here](https://github.com/dimagi/commcare-hq/blob/d58c94383e9883ba67c7e6355e9cd213f14617a5/corehq/messaging/scheduling/forms.py#L3656)). 


### knockout.js...knocked me out a little
The dedupe feature is not written using crispy forms, so reusing the existing filter criteria crispy form required "injecting" that form into a child element of the dedupe form (see [here](https://github.com/dimagi/commcare-hq/blob/0d1de55a05d6f357e2850a3130125e1c9ad9acaf/corehq/apps/data_interfaces/templates/data_interfaces/edit_deduplication_rule.html#L83)).  
Because we're now using a crispy form inside another form, the outer form's ko binding does not work. More specifically, knockout.js throws a "Multiple bindings to same element" error, which, from what I can gather, simply means we can't bind to an element which already had a `.koApplyBindings` called on it ([this](https://github.com/dimagi/commcare-hq/blob/03e7f7035235a667d5095c8af6caea9ac4ab79b2/corehq/apps/data_interfaces/static/data_interfaces/js/case_rule_criteria.js#L305) binding is made on the new "injected" crispy filter form before [this](https://github.com/dimagi/commcare-hq/blob/0d1de55a05d6f357e2850a3130125e1c9ad9acaf/corehq/apps/data_interfaces/static/data_interfaces/js/case_dedupe_main.js#L91) binding is made on the "outer" dedupe form). 

After a bit of playing around I got it to work (see embarassing [hack](https://github.com/dimagi/commcare-hq/blob/0d1de55a05d6f357e2850a3130125e1c9ad9acaf/corehq/apps/data_interfaces/static/data_interfaces/js/case_dedupe_main.js#L89)) by removing the element which is first being bound to from the "outer" form till after the "outer" form's binding has been made, after which the first element is simply appended again (feels like duct-taping the problem). I don't like it...not a bit, but it works. I am, however, looking into more elegant ways to deal with this matter. 

## Feature Flag
[Deduplication](https://confluence.dimagi.com/pages/viewpage.action?spaceKey=saas&title=Surfacing+Case+Duplicates+in+CommCare)

## Safety Assurance

### Safety story
1) Some local testing
2) PR includes automated tests
3) Did some testing on staging

### Automated test coverage
- New location criterion tests
- Tests for constructing the query used for finding duplicate cases
- Tests for dedupe rule execution with criteria

### QA
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-4119)

### Migrations
- [ ] The migrations in this code can be safely applied first independently of the code

It seems [this](https://github.com/dimagi/commcare-hq/blob/61b22b78130eac973c0a86e94fbca727bbd1a9e4/corehq/apps/data_interfaces/migrations/0026_auto_20220420_1301.py#L27) migration is dependant on [this](https://github.com/dimagi/commcare-hq/blob/6bfbad00502f6e0800f913492e61fb398e39bd72/corehq/apps/data_interfaces/models.py#L673) code existing.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Charl1996 committed:
Add Case Filter section to dedupe rule UI

Charl1996 committed:
Add locations filter option

Charl1996 committed:
Add LocationMatchDefinition criteria

Charl1996 committed:
Add migration for new model

Charl1996 committed:
Filter dedupe cases based on rule criteria

Charl1996 committed:
Add some tests

Charl1996 committed:
Add option to include_child_locations

Charl1996 committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 committed:
Add allow_custom_filter method

Charl1996 committed:
Consider user role when retrieving user_locations

Charl1996 committed:
Remove redundant fields

Charl1996 committed:
Add new model to some important places

Charl1996 committed:
Add migrations.lock file

Charl1996 committed:
Stickler standards

Charl1996 committed:
Fix test

Charl1996 committed:
Format dedupe error message better

Charl1996 committed:
Check case location_id in LocationFilterDefinition match

Charl1996 committed:
Fix LocationFilterDefinition matches method to correctly check for user ownership at location

Charl1996 committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 commented:
Created this PR after closing [this](https://github.com/dimagi/commcare-hq/pull/31537) one.

Charl1996 commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -1150,21 +1171,38 @@ def post(self, request, *args, **kwargs):
+            # Is there a good way to check if the criteria has changed without
+            # going over each and every value?
+            cases_filter_form.save_criteria(self.rule, save_meta=False)
+            filter_criteria_updated = True
```
`form.save_criteria()` overwrites the criteria definitions by wiping the current ones and reconstructing new ones based on the form cleaned_data...but this leaves us without knowing if there ever was a change made...I was thinking of rewriting that method (or writing a new one) so that we can know if any criteria was actually updated, but going through the current criteria and comparing it to the new ones seems...sub-optimal...but I could be mistaken.

Any thoughts?

Charl1996 committed:
Remove redundant case search methods

sravfeyn commented on corehq/apps/data_interfaces/deduplication.py:
```diff
@@ -4,16 +4,63 @@
 
-def find_duplicate_case_ids(domain, case, case_properties, include_closed=False, match_type="ALL"):
-    es = CaseSearchES().domain(domain).size(DUPLICATE_LIMIT).case_type(case.type)
+def _get_es_query(domain, case, case_filter_criteria=[]):
```
May be this could be renamed to better indicate kind of the query?

sravfeyn commented on corehq/apps/data_interfaces/forms.py:
```diff
@@ -422,6 +445,26 @@ def __init__(self, domain, *args, **kwargs):
+        if self.couch_user is None:
+            return []
+
+        if self.couch_user.is_domain_admin(self.domain):
```
I am surprised that there isn't already a method to fetch these, have you checked if you could reuse anything?

Charl1996 replied:
@sravfeyn I see this is the way it's checked in a bunch of other places. I haven't found another cleaner way yet.

sravfeyn commented on corehq/apps/data_interfaces/forms.py:
```diff
@@ -567,19 +610,34 @@ def clean_property_match_definitions(self):
+
+        if value:
+            location_def = value[0]
+            if not location_def['include_child_locations']:
```
Do you need to handle `KeyError`?

Charl1996 replied:
Good point.

I'm thinking along the lines of simply doing this: 
```suggestion
            if not location_def.get('include_child_locations'):
```

sravfeyn commented on corehq/apps/data_interfaces/models.py:
```diff
@@ -673,6 +683,41 @@ def matches(self, case, now):
         return False
 
 
+class LocationFilterDefinition(CaseRuleCriteriaDefinition):
```
Should this model be added to the domain dump/load models?

sravfeyn replied:
Ah, I see it's already added

Charl1996 committed:
Fix query construction

Charl1996 committed:
Remove redundant caching and rename function

Charl1996 committed:
Address PR comments

Charl1996 committed:
Mr. Stickler has spoken

Charl1996 committed:
Fix es issues in test

esoergel reviewed (commented):
Left some feedback on the case search and locations logic, didn't do a full review otherwise

Charl1996 committed:
Remove LocationFilterDefinition on location delete

esoergel commented on corehq/apps/locations/models.py:
```diff
@@ -422,6 +422,8 @@ def delete(self, *args, **kwargs):
         for loc in to_delete:
             publish_location_saved(loc.domain, loc.location_id, is_deletion=True)
 
+        delete_locations_filter_criteria.delay([loc.location_id for loc in to_delete])
```
Since this is a SQL model, you can also set it up to use a cascading delete like here:
https://github.com/dimagi/commcare-hq/blob/bee1fd160139f21ef3a629268a0d44f3a0e02afa/corehq/apps/app_manager/models.py#L6047-L6047
That way PostgreSQL will handle it for us and we don't need to worry about celery tasks.

Charl1996 replied:
@esoergel I updated the code, but decided to take another approach after I have rethinked it a bit. This time I remove the rule completely which depends on the location, since the location-dependant rule does not make sense after a location is deleted. If I only remove the definition from the rule, the rule might affect a wrong set of cases since the criteria changed.
As such, I renamed the function and kept it as a celery task to be executed on location delete.

gherceg committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

gherceg commented on corehq/apps/es/case_search.py:
```diff
@@ -283,10 +283,7 @@ def _case_property_not_set(case_property_name):
-                queries.match_all(),
-                filters.term('{}.key.exact'.format(CASE_PROPERTIES_PATH), case_property_name),
-            )
+            filters.term(PROPERTY_KEY, case_property_name),
```
fyi I resolved a conflict with master by mixing @esoergel's recent changes with the change you have in this PR @Charl1996. Just wanted to draw attention to this in case it needs to change.

Charl1996 committed:
Fix test and delet rule after location delete

Charl1996 committed:
Update migrations

Charl1996 committed:
Ran make translations

Charl1996 committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 committed:
Igonore coverage-report

Charl1996 committed:
Ran make translations

Charl1996 committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 committed:
Fix Stickler issue

Charl1996 committed:
Minimise migration files

Charl1996 committed:
Update migration lock file

Charl1996 committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 committed:
Fix ucr filter form cleaning function

Charl1996 committed:
Update migrations.lock

Charl1996 committed:
Fix translations

kaapstorm committed:
Location functions

kaapstorm committed:
Tests

kaapstorm committed:
Use functions

kaapstorm committed:
isort

kaapstorm commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -1150,21 +1171,38 @@ def post(self, request, *args, **kwargs):
-            messages.error(request, error_message + "; ".join(errors))
+            messages.error(
+                request, _('Deduplication rule not saved due to the following issues: %s') %
+                '<ul><li>{}</li></ul>'.format('</li><li>'.join(errors)),
```
:thinking: On second thoughts, maybe we can't. HTML in Python just feels awkward.

Charl1996 commented on corehq/apps/data_interfaces/deduplication.py:
```diff
@@ -4,16 +4,63 @@
 
-def find_duplicate_case_ids(domain, case, case_properties, include_closed=False, match_type="ALL"):
-    es = CaseSearchES().domain(domain).size(DUPLICATE_LIMIT).case_type(case.type)
+def _get_es_filtered_case_query(domain, case, case_filter_criteria=[]):
+    # Import here to avoid circular import error
+    from corehq.apps.data_interfaces.models import (
+        MatchPropertyDefinition,
+        LocationFilterDefinition
+    )
```
Oh, interesting caveat!

Charl1996 commented on corehq/apps/data_interfaces/forms.py:
```diff
@@ -862,3 +930,78 @@ def save_actions(self, rule):
+        return False
+
+    def __init__(self, domain, *args, **kwargs):
+        couch_user = kwargs.get('couch_user', {})
+        kwargs['is_system_admin'] = couch_user.is_superuser if couch_user else False
+        super(DedupeCaseFilterForm, self).__init__(domain, *args, **kwargs)
```
Mmm, you're definitely right. I see the confusion.

Charl1996 committed:
Make defaults non-mutable

Charl1996 committed:
Merge branch 'nh/location_funcs' into cs/SC-1945-ability-to-add-filter-to-rules

kaapstorm reviewed (approved):
Thanks for these changes @Charl1996

Charl1996 commented on corehq/apps/locations/models.py:
```diff
@@ -796,3 +796,23 @@ def get_case_sharing_groups_for_locations(locations, for_user_id=None):
+    return SQLLocation.active_objects.filter(domain=domain)
+
+
+def get_accessible_locations(domain, couch_user):
```
Oh, I was not aware of its existence. Thanks for the reference! I'll update.

Charl1996 committed:
Remove redundant function

Charl1996 commented on corehq/apps/locations/tasks.py:
```diff
@@ -124,6 +125,17 @@ def update_users_at_locations(domain, location_ids, supply_point_ids, ancestor_i
+            rule = criteria.rule
+            rule.delete_criteria()
+            rule.delete_actions()
+            rule.delete()
```
@esoergel 
> Don't rules commonly have multiple criteria?

Are you asking whether rules share criteria? If so, the answer is no. You can see [here](https://github.com/dimagi/commcare-hq/blob/db36c7d617756ea5b27e1c360769ead347bd5e4b/corehq/apps/data_interfaces/forms.py#L657) when the criteria is [added to the rule](https://github.com/dimagi/commcare-hq/blob/c8eb0c252525806f28f158cf23b7c08e0f7e562c/corehq/apps/data_interfaces/views.py#L1020), the existing criteria is removed completely and recreated, meaning that the criteria on any rule is only used on that rule and not shared with any other rule. 

It may also be worth mentioning that any rule can only have exactly one location filter (at least at the moment; OR logic is not implemented for dedupe yet), so when we're iterating over the location_definitions and looping over their respective criteria, the rule returned when we call `criteria.rule` will always be unique for every iteration.

As for the tests, I'll add more verbose testing :+1:

esoergel replied:
I was asking whether there can be more than one criteria object pointing to the same rule.

> so when we're iterating over the location_definitions and looping over their respective criteria, the rule returned when we call criteria.rule will always be unique for every iteration.

Perfect, this is what I was wondering about - that sounds fine then.

Charl1996 committed:
Add some more tests

orangejenny commented:
> I am, however, looking into more elegant ways to deal with this matter.

I haven't looked at the code closely, but based on your description, you might find the [stopBinding](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js#L614-L639) useful.

Charl1996 committed:
Check if user exist when pulling user locations

Charl1996 commented:
@orangejenny 
> I haven't looked at your code closely, but based on your description of the issue, you might find the [stopBinding](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js#L614-L639) useful.

I remember having tried something like that after googling a bit and it did not work, but it's probably the case that I just missed something back then. I'll try again, thanks!

Charl1996 committed:
Ignore unimportant hard fail

Charl1996 committed:
Merge branch 'cs/SC-1945-ability-to-add-filter-to-rules' of github.com:dimagi/commcare-hq into jls/dd-in-rules-ui+cs/SC-1945-ability-to-add-filter-to-rules

orangejenny committed:
Added criteria init to case_dedup_main

orangejenny committed:
Added autocompletes to criteria and actions sections of dedupe rules page

orangejenny committed:
Revert "Added autocompletes to criteria and actions sections of dedupe rules page"

This reverts commit 5e9344473b04b8d5a6f7e55dfbdeafb2f2c0d40e.

orangejenny committed:
Merge branch 'jls/dd-in-rules-ui' into jls/dd-in-rules-ui+cs/SC-1945-ability-to-add-filter-to-rules

orangejenny committed:
Updated autocompletes to use new component

orangejenny committed:
Merge branch 'jls/dd-in-rules-ui' into jls/dd-in-rules-ui+cs/SC-1945-ability-to-add-filter-to-rules

orangejenny commented:
Merged the conflict branch from this and https://github.com/dimagi/commcare-hq/pull/31854 which is now merged to master.

dannyroberts commented:
Seems like somethings up with the history / diff in this PR now. (Not sure if it's a side effect of the merge you just did @orangejenny or has been there for some time.) The concrete thing I notice is that the diff (and commits) from https://github.com/dimagi/commcare-hq/pull/31794 are showing up in this PR too.

orangejenny committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 committed:
Revert to 4440b10

Charl1996 committed:
Merge branch 'master' into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 committed:
Merge branch 'cs/SC-1945-ability-to-add-filter-to-rules' of github.com:dimagi/commcare-hq into cs/SC-1945-ability-to-add-filter-to-rules

Charl1996 committed:
Include users as case owners when executing case search

Charl1996 committed:
Add test for closed cases

Charl1996 committed:
Fix test

Charl1996 committed:
Remove duplicate definition

orangejenny commented:
This popped up on staging and looks possibly related to this PR, can you check it out? https://sentry.io/organizations/dimagi/issues/3262300213/

dannyroberts commented:
Reviewing our August PRs, I saw this one was merged without a final approval (and with what appear to be substantial changes since the approval). @Charl1996 can you note if there's any additional context to that?

dannyroberts commented:
Reviewing our August PRs, I saw this one was merged without a final approval (and with what appear to be substantial changes since the approval). @Charl1996 can you note if there's any additional context to that?

Charl1996 commented:
@dannyroberts 

Yeah, this is one of my embarassing blunders. I was hasty and this PR was open for quite a while so I thought the changes where insignificant enough to just bypass review...(I cringe just by typing this)

> what appear to be substantial changes since the approval

Not really too substantial, but yes, substantial enough. I think the most significant one is the [Revert to](https://github.com/dimagi/commcare-hq/pull/31629/commits/4e387dcb05ca85fe0fefed4e7ead20964cccb232) https://github.com/dimagi/commcare-hq/commit/4440b1078121a8e85c31e196a0fcea1fe6763439 where I reverted back to a certain commit after there was some conflict between my branch and one of Jenny's. Somehow the code got all tangled up, so I decided to revert to a place I knew it worked and updated the branch with master (this was now after Jenny's work has gone to master). 

Apart from the above it was only small stuff, like removing redundant code or adding some tests / fixing QA bugs (checking if a user exist or so). In my head I thought the testing additions wasn't really too important to be reviewed, but after a chat with Daniel about this premature merge I realised that my mindset about this was completely off the rails.

I'll also just note here for future readers that all the changes (34 files) that was merged are intended changes by me. Nothing that is included in this merge is unintentional or "slipped" through somehow with all the commit shenanigans.

dannyroberts commented:
Okay, thanks for the response. Sounds like you retro'd the mistake already at the time and the team used it as a learning opportunity, which is a good response. I've also reviewed the commits since the last approval and approve those retroactively ✅.
